### PR TITLE
 [vcpkg_find_acquire_program] Enable find_acquire(PKGCONFIG)

### PR DESCRIPTION
**Pull request by ras0219**
This PR enables `vcpkg_find_acquire_program(PKGCONFIG)` so we do not need to rely on msys2 for this common component.

This uses a Win32-native build of `pkg-config`, which handles Win32 paths without cygwin/msys conversion and uses semicolon-delimited lists in path variables.

### CODE REVIEW

**Comment by Neumann-A** [comment]
Can MS not open a repo called vcpkg-buildtools or similiar which hosts all the tools on github? Pkg-config can easily be built by vcpkg. See #11523

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 1]
```diff
@@ -349,6 +349,51 @@ function(vcpkg_find_acquire_program VAR)
...
     set(ARCHIVE "aria2-1.34.0-win-32bit-build1.zip")
     set(HASH 2a5480d503ac6e8203040c7e516a3395028520da05d0ebf3a2d56d5d24ba5d17630e8f318dd4e3cc2094cc4668b90108fb58e8b986b1ffebd429995058063c27)
+  elseif(VAR MATCHES "PKGCONFIG")
+    set(PROGNAME pkg-config)
+    if(CMAKE_HOST_WIN32)
+      set(PROG_PATH_SUBDIR "${DOWNLOADS}/tools/${PROGNAME}/0.26-1")
+      set(PKGCONFIG "${PROG_PATH_SUBDIR}/bin/pkg-config.exe")
+      if(NOT EXISTS "${PKGCONFIG}")
+        vcpkg_download_distfile(PKGCONFIG_ARCHIVE
+          URLS "http://ftp.gnome.org/pub/gnome/binaries/win32/dependencies/pkg-config_0.26-1_win32.zip"
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:359]: That is a rather old version of pkg-config(from 2011?). I dont know if it handles everything correctly. Current major is 0.29

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 2]
```diff
@@ -349,6 +349,51 @@ function(vcpkg_find_acquire_program VAR)
...
+          COMMAND ${CMAKE_COMMAND} -E tar xzf ${PKGCONFIG_ARCHIVE}
+          WORKING_DIRECTORY ${PROG_PATH_SUBDIR}
+        )
+      endif()
+      set(${VAR} "${${VAR}}" PARENT_SCOPE)
+      return()
+    else()
+      set(BREW_PACKAGE_NAME pkg-config)
+      set(APT_PACKAGE_NAME pkg-config)
+      set(PATHS "/bin" "/usr/bin" "/usr/local/Cellar/pkg-config/0.29.2_3")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:395]: the last one should be `/usr/local/bin` instead of cellar for osx

Comment (Neumann-A) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:395]:
```suggestion
      set(PATHS "/bin" "/usr/bin" "/usr/local/bin")
```

Comment (ras0219) [reply scripts/cmake/vcpkg_find_acquire_program.cmake]: Thanks

**Comment by Voskrese** [comment]
`https://packages.msys2.org/base/mingw-w64-pkg-config`

_____





_____

`https://repo.msys2.org/mingw/x86_64/mingw-w64-x86_64-pkg-config-0.29.2-1-any.pkg.tar.xz`

```
pkg-config.exe --help
Usage:
  pkg-config.exe [OPTION...]

Help Options:
  -h, --help                              Show help options

Application Options:
  --version                               output version of pkg-config
  --modversion                            output version for package
  --atleast-pkgconfig-version=VERSION     require given version of pkg-config
  --libs                                  output all linker flags
  --static                                output linker flags for static linking
  --short-errors                          print short errors
  --libs-only-l                           output -l flags
  --libs-only-other                       output other libs (e.g. -pthread)
  --libs-only-L                           output -L flags
  --cflags                                output all pre-processor and compiler flags
  --cflags-only-I                         output -I flags
  --cflags-only-other                     output cflags not covered by the cflags-only-I option
  --variable=NAME                         get the value of variable named NAME
  --define-variable=NAME=VALUE            set variable NAME to VALUE
  --exists                                return 0 if the module(s) exist
  --print-variables                       output list of variables defined by the module
  --uninstalled                           return 0 if the uninstalled version of one or more module(s) or their dependencies will be used
  --atleast-version=VERSION               return 0 if the module is at least version VERSION
  --exact-version=VERSION                 return 0 if the module is at exactly version VERSION
  --max-version=VERSION                   return 0 if the module is at no newer than version VERSION
  --list-all                              list all known packages
  --debug                                 show verbose debug information
  --print-errors                          show verbose information about missing or conflicting packages (default unless --exists or --atleast/exact/max-version given on the command line)
  --silence-errors                        be silent about errors (default when --exists or --atleast/exact/max-version given on the command line)
  --errors-to-stdout                      print errors from --print-errors to stdout not stderr
  --print-provides                        print which packages the package provides
  --print-requires                        print which packages the package requires
  --print-requires-private                print which packages the package requires for static linking
  --validate                              validate a package's .pc file
  --define-prefix                         try to override the value of prefix for each .pc file found with a guesstimated value based on the location of the .pc file
  --dont-define-prefix                    don't try to override the value of prefix for each .pc file found with a guesstimated value based on the location of the .pc file
  --prefix-variable=PREFIX                set the name of the variable that pkg-config automatically sets
  --msvc-syntax                           output -l and -L flags for the Microsoft compiler (cl)
```

**Comment by Voskrese** [comment]
@Neumann-A 
add to `vcpkg_fixup_pkgconfig`

```
vcpkg_acquire_msys(MSYS_ROOT PACKAGES pkg-config)
find_program(PKGCONFIG_EXECUTABLE NAMES pkg-config HINTS ${MSYS_ROOT} PATH_SUFFIXES "usr/bin" NO_DEFAULT_PATH)
mark_as_advanced(FORCE PKGCONFIG_EXECUTABLE)
set(PKGCONFIG ${PKGCONFIG_EXECUTABLE})
```

**Comment by ras0219** [comment]
Thanks for the link @Voskrese!

Based on my testing, the mingw32 copy of pkg-config only requires one additional dep (libwinpthread). This should give us the best of both worlds: up-to-date pkg-config as well as freedom from `pacman` instabilities.

If this works, I believe this is a technique we can apply across the ecosystem to remove dependence on msys pacman.

**Comment by ras0219** [comment]
The lapack errors are unrelated; in CI:
```
-- The Fortran compiler identification is unknown
-- The C compiler identification is unknown
CMake Error at CMakeLists.txt:3 (project):
  The CMAKE_Fortran_COMPILER:

    D:/downloads/tools/msys2/msys64/mingw32/bin/gfortran.exe

  is not a full path to an existing compiler tool.

  Tell CMake where to find the compiler by setting either the environment
  variable "FC" or the CMake cache entry CMAKE_Fortran_COMPILER to the full
  path to the compiler, or to the compiler name if it is in the PATH.


CMake Error at CMakeLists.txt:3 (project):
  The CMAKE_C_COMPILER:

    D:/downloads/tools/msys2/msys64/mingw32/bin/gcc.exe

  is not a full path to an existing compiler tool.

  Tell CMake where to find the compiler by setting either the environment
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
  the compiler, or to the compiler name if it is in the PATH.


-- Configuring incomplete, errors occurred!
```

when trying to repro locally, vcpkg selected ninja from my VS instance:
```
.... "-DCMAKE_MAKE_PROGRAM=C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" ...
CMake Error at CMakeLists.txt:3 (project):
  The Ninja generator does not support Fortran using Ninja version

    1.8.2

  due to lack of required features.  Ninja 1.10 or higher is required.


-- Configuring incomplete, errors occurred!
```

lapack-reference failures are the causes of `x64-windows`, `x64-windows-static`, `x86-windows`, and `x64-uwp` failing.

File: ports/geogram/portfile.cmake [hunk 3]
```diff
@@ -2,7 +2,7 @@ set(GEOGRAM_VERSION 1.7.5)
 
 vcpkg_download_distfile(ARCHIVE
     URLS "https://gforge.inria.fr/frs/download.php/file/38314/geogram_${GEOGRAM_VERSION}.tar.gz"
-    FILENAME "geogram_${GEOGRAM_VERSION}.tar.gz"
+    FILENAME "geogram_${GEOGRAM_VERSION}_47dcbb8.tar.gz"
```
Comment (BillyONeal) [inline ports/geogram/portfile.cmake:5]: Why did this change?

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 4]
```diff
@@ -123,20 +112,14 @@ function(vcpkg_configure_meson)
...
             set(ENV{PKG_CONFIG} "${PKGCONFIG} --define-variable=prefix=${PKGCONFIG_PREFIX}")
         endif()
         set(PKGCONFIG_INSTALLED_DIR "${CURRENT_INSTALLED_DIR}/debug/lib/pkgconfig/")
-        if(WIN32)
-            string(REGEX REPLACE "([a-zA-Z]):/" "/\\1/" PKGCONFIG_INSTALLED_DIR "${PKGCONFIG_INSTALLED_DIR}")
-        endif()
         if(ENV{PKG_CONFIG_PATH})
             set(BACKUP_ENV_PKG_CONFIG_PATH_DEBUG $ENV{PKG_CONFIG_PATH})
-            set(ENV{PKG_CONFIG_PATH} "${PKGCONFIG_INSTALLED_DIR}: ${PKGCONFIG_SHARE_DIR}: $ENV{PKG_CONFIG_PATH}")
+            set(ENV{PKG_CONFIG_PATH} "${PKGCONFIG_INSTALLED_DIR}${VCPKG_HOST_PATH_SEPARATOR}${PKGCONFIG_SHARE_DIR}${VCPKG_HOST_PATH_SEPARATOR}$ENV{PKG_CONFIG_PATH}")
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_configure_meson.cmake:120]: Stupid question: can this just be / ?

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 5]
```diff
@@ -378,6 +378,47 @@ function(vcpkg_find_acquire_program VAR)
...
     set(ARCHIVE "aria2-1.34.0-win-32bit-build1.zip")
     set(HASH 2a5480d503ac6e8203040c7e516a3395028520da05d0ebf3a2d56d5d24ba5d17630e8f318dd4e3cc2094cc4668b90108fb58e8b986b1ffebd429995058063c27)
+  elseif(VAR MATCHES "PKGCONFIG")
+    set(PROGNAME pkg-config)
+    if(ENV{PKG_CONFIG})
+      debug_message(STATUS "PKG_CONFIG found in ENV! Using $ENV{PKG_CONFIG}")
+      set(PKGCONFIG $ENV{PKG_CONFIG} PARENT_SCOPE)
+      return()
+    elseif(CMAKE_HOST_WIN32)
+      set(PROG_PATH_SUBDIR "${DOWNLOADS}/tools/${PROGNAME}/0.29.2-1")
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:388]: It seems very strange to have `${PROGNAME}` but then a specific hardcoded version string?

File: ports/geogram/portfile.cmake (hunk 3 above)
Comment (ras0219) [reply ports/geogram/portfile.cmake:5]: On one of the CI machines, there was a extant `geogram_${GEOGRAM_VERSION}.tar.gz` file with the previous version's hash. 

Renaming the file resolved the conflict.

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 4 above)
Comment (ras0219) [reply scripts/cmake/vcpkg_configure_meson.cmake:120]: Ah, sorry for the confusion -- `PATH` here refers to the environment variable, so it's either `;` or `:`.

File: ports/geogram/portfile.cmake (hunk 3 above)
Comment (BillyONeal) [reply ports/geogram/portfile.cmake:5]: I see; it just looks odd to have that variable `GEOGRAM_VERSION` but this be a hardcoded constant. Not necessarily worth resetting CI over but it's confusing nonetheless

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 4 above)
Comment (BillyONeal) [reply scripts/cmake/vcpkg_configure_meson.cmake:120]: Bwahahahaha OK

File: scripts/cmake/vcpkg_find_acquire_program.cmake (hunk 5 above)
Comment (ras0219) [reply scripts/cmake/vcpkg_find_acquire_program.cmake:388]: That's a good point. I can change it, but that would invoke a total rebuild of the tree.

**Comment by BillyONeal** [comment]
Thanks for your contribution!

**Pull request merged**